doc/reference/common-weaknesses/hard-limits.rst

```diff
@@ -0,0 +1,72 @@
+- *Soft limits*: these are limits which *can* be breached, but where there is a significant disincentive to do so. One way of implementing a soft limit is to have sharply increasing costs to using the resource beyond the soft limit.
+
+Hard limits are clear, easy to specify, and provide hard guarantees for the protocol, but they have the disadvantage that there is no way to evade the limit.
+This means that there is a discontinuity at the limit: beforehand you can always do more by paying more, but after the limit there is nothing you can do.
+
+Most of the limits for resources on Cardano are hard limits:
+
+- Transaction size has a hard limit.
+- Block size has a hard limit.
+- UTXO size has a hard limit.
+- Script execution units have a hard limit.
+
+If an application *requires* a transaction that exceeds one of these limits, then the application will be stuck unless the limit is increased or removed.
+This is most common when scripts are involved, since a script can require a very particular shape of transaction, regardless of whether this exceeds limits.
+
+Examples:
+
+- A script requires providing a datum which is extremely large and exceeds the transaction size limit.
+- A script needs more execution units than the limit under unusual circumstances, rendering the output which it locks unspendable.
+- A script requires creating a single output containing a very large amount of tokens, which exceeds the output size limit.
+
+Risks
+~~~~~
+
+This is typically an issue for applications that work with user-supplied data, or data that can grow in an unbounded way over time.
+This can result in data which itself becomes large, or which requires a large amount of resources to process.
+
+For example:
+
+- Managing an arbitrary collection of assets (unbounded over time).
+- Allowing user-specified payloads in datums (user-supplied unbounded data).
+
+Script size should not itself be a risk (since scripts and their sizes should generally be known ahead of time), but large scripts can reduce the amount of space available for other uses, heightening the risk of hitting a limit.
+
+Solutions
+~~~~~~~~~
+
+In the long run, hard limits may be increased, removed, or turned into soft limits.
+
+In the mean time, there are some approaches that developers can use to reduce the risk.
+
+Careful testing
+---------------
+
+It is important to test as many of the execution paths of your application as possible.
+This is important for correctness, but also to ensure that there are not unexpected cases where script resource usage spikes.
+
+Bounding data usage
+-------------------
+
+Carefully consider whether your application may rely on unbounded data, and try to avoid that.
+For example, if your application needs to manage a large quantity of assets, try to split them across multiple UTXOs instead of relying on a single UTXO to hold them all.
+
```

Comment:
Read this gave me the idea that we should perhaps give the plutus user some feedback when writing a plutus script, based on the current hard limits of the chain:

e.g. 
```
current cardano transaction size = X KB
this plutus script = Y KB
size left for datums = X-Y KB
```

also do u have a suggestion of how the user can "bound" their Datums? Some example offline code would help. Can we suggest something to the users for checking their Datums? `Flat.size` or `serialisedSize` or something else?

Reply:
Yeah, we could do this in `plutus-apps`, but we don't know all these numbers here...

There isn't really a general solution to bounding your data usage, you just have to think about it! And notice if you're doing something that commits you to using an unbounded amount of data.